drivers/regulator/regulator_axp2101.c

```diff
@@ -0,0 +1,573 @@
+	LOG_INST_DBG(config->log, "[0x%02x]=0x%02x mask=0x%02x", config->desc->enable_reg,
+		     config->desc->enable_val, config->desc->enable_mask);
+
+	/* special case for LDOIO0, which is multiplexed with GPIO0 */
+	ret = i2c_reg_update_byte_dt(&config->i2c, config->desc->enable_reg,
+				     config->desc->enable_mask, config->desc->enable_val);
+
+	if (ret != 0) {
+		LOG_INST_ERR(config->log, "Failed to enable regulator");
+	}
+
+	return ret;
+}
+
+static int axp2101_disable(const struct device *dev)
+{
+	const struct regulator_axp2101_config *config = dev->config;
+	int ret;
+
+	LOG_INST_DBG(config->log, "Disabling regulator");
+	LOG_INST_DBG(config->log, "[0x%02x]=0 mask=0x%x", config->desc->enable_reg,
+		     config->desc->enable_mask);
+
+	/* special case for LDOIO0, which is multiplexed with GPIO0 */
+	ret = i2c_reg_update_byte_dt(&config->i2c, config->desc->enable_reg,
+				     config->desc->enable_mask, 0u);
+
+	if (ret != 0) {
+		LOG_INST_ERR(config->log, "Failed to disable regulator");
+	}
+
+	return ret;
+}
+
+static unsigned int axp2101_count_voltages(const struct device *dev)
+{
+	const struct regulator_axp2101_config *config = dev->config;
+
+	return linear_range_group_values_count(config->desc->ranges, config->desc->num_ranges);
+}
+
+static int axp2101_list_voltage(const struct device *dev, unsigned int idx, int32_t *volt_uv)
+{
+	const struct regulator_axp2101_config *config = dev->config;
+
+	return linear_range_group_get_value(config->desc->ranges, config->desc->num_ranges, idx,
+					    volt_uv);
+}
+
+static int axp2101_set_voltage(const struct device *dev, int32_t min_uv, int32_t max_uv)
+{
+	const struct regulator_axp2101_config *config = dev->config;
+	uint16_t idx;
+	int ret;
+
+	LOG_INST_INF(config->log, "voltage = [min=%d, max=%d]", min_uv, max_uv);
+
+	/* set voltage */
+	ret = linear_range_group_get_win_index(config->desc->ranges, config->desc->num_ranges,
+					       min_uv, max_uv, &idx);
+	if (ret != 0) {
+		LOG_INST_ERR(config->log, "No voltage range window could be detected");
+		return ret;
+	}
+
+	idx <<= config->desc->vsel_bitpos;
+
+	LOG_INST_INF(config->log, "[0x%x]=0x%x mask=0x%x", config->desc->vsel_reg, idx,
+		     config->desc->vsel_mask);
+	ret = i2c_reg_update_byte_dt(&config->i2c, config->desc->vsel_reg, config->desc->vsel_mask,
+				     (uint8_t)idx);
+	if (ret != 0) {
+		LOG_INST_ERR(config->log, "Failed to set regulator voltage");
+	}
+
+	return ret;
+}
+
+static int axp2101_get_voltage(const struct device *dev, int32_t *volt_uv)
+{
+	const struct regulator_axp2101_config *config = dev->config;
+	int ret;
+	uint8_t raw_reg;
+
+	/* read voltage */
+	ret = i2c_reg_read_byte_dt(&config->i2c, config->desc->vsel_reg, &raw_reg);
+	LOG_INST_INF(config->log, "voltage reg=0x%x", raw_reg);
+	if (ret != 0) {
+		return ret;
+	}
+
+	raw_reg = (raw_reg & config->desc->vsel_mask) >> config->desc->vsel_bitpos;
+
+	ret = linear_range_group_get_value(config->desc->ranges, config->desc->num_ranges, raw_reg,
```

Comment:
This seems to be incorrect for DCDC2, DCDC3, and DCDC4.
`linear_range_group_get_value` does not handle an array.